kmk/firmware.py

```diff
@@ -75,12 +83,83 @@ def _send_key(self, key):
             self._state.force_keycode_up(key)
             self._send_hid()
 
+    def _handle_update(self, update):
+        if update is not None:
+            self._state.matrix_changed(
+                update[0],
+                update[1],
+                update[2],
+            )
+
+            if self._state.hid_pending:
+                self._send_hid()
+
+            if self.debug_enabled:
+                print('New State: {}'.format(self._state._to_dict()))
+
+        self._state.process_timeouts()
+
+        for key in self._state.pending_keys:
+            self._send_key(key)
+            self._state.pending_key_handled()
+
+        if self._state.macro_pending:
+            for key in self._state.macro_pending(self):
+                self._send_key(key)
+
+            self._state.resolve_macro()
+
+    def _send_to_master(self, update):
+        if self.uart is not None:
+
+            if self.split_master_left:
+                update[1] += self.split_offsets[update[0]]
+
+            self.uart.write(update)
+
+    def _receive_from_slave(self):
+        if self.uart is not None and self.uart.in_waiting > 0:
+            update = bytearray(self.uart.read(3))
+            if self.split_master_left:
+                update[1] -= self.split_offsets[update[0]]
+            return update
+
+        return None
+
+    def _master_half(self):
```

Comment:
Cache this to a variable rather than using a function here, maybe? Saves some cycles.

Reply:
Same as before, bluetooth functionality will be different, and therefor this was left open ended for changes.

Reply:
Caching now.